{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bioinformatics tools to predict protein properties from sequence:\n",
    "\n",
    "The first part of this tutorial covers the assignment of protein properties from their structural informations and the second part covers the prediction of properties from sequence for proteins without the structure (about 98% of all proteins):\n",
    "\n",
    "1. Assignment of protein properties (for proteins with structural information)\n",
    "2. prediction of protein properties \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Assignment of protein properties (for proteins with structural information)\n",
    "There are now well over 150 million protein sequences available from Uniprot, and For only a small fraction of these (about 2%) structural information is available from the Protein Data Bank.\n",
    "For this fraction of proteins PDBe-KB  can be used. For the rest we can predict the properties from sequence.\n",
    "\n",
    "## PDBe-KB\n",
    "\n",
    "PDBe-KB (Protein Data Bank in Europe- Knowledge Base) Collating functional annotations and predictions for structure data in the PDB archive. The following steps are some information the PDBe-KB REST API can give:\n",
    "\n",
    "1.Electron density:\n",
    "This call details the statistics for electron density.\n",
    "\n",
    "2.Functional annotation:\n",
    "It provides functional annotation of all ligands, i.e. 'bound' molecules that are not waters if available.\n",
    "\n",
    "3.EBI Assembly:\n",
    "It provides information for each assembly of a given PDB ID. This information is broken down at the entity level for each assembly. The information given includes the molecule name, type and class, the chains where the molecule occur, and the number of copies of each entity in the assembly.\n",
    "\n",
    "4.EBI Ratio of observed residues:\n",
    "It provides the ratio of observed residues for each chain in each molecule (or entity in mmcif-speak) of a pdb entry. The list of chains within an entity is sorted by observed_ratio (descending order), partial_ratio (ascending order), and number_residues (descending order).\n",
    "\n",
    "5.Binding sites:\n",
    "It provides details on binding sites in the entry as per STRUCT_SITE records in PDB files (or mmcif equivalent thereof), such as ligand, residues in the site, description of the site, etc.\n",
    "\n",
    "6.Secondary structure:\n",
    "It provides details about residue ranges of regular secondary structure (alpha helices and beta strands) found in protein chains of the entry. For strands, sheet id can be used to identify a beta sheet.\n",
    "\n",
    "7.Observed ranges:\n",
    "It provides observed ranges, i.e. segments of structural coverage, of polymeric molecules that are modelled fully or partly.Below is an example of one of these properties and it will be the same for the others.\n",
    "\n",
    "You can also access the [PDBe-kb server](https://www.ebi.ac.uk/pdbe/pdbe-kb) directly.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2n4p, 5mdi, 5mrg, 5x4f, 6b1g\n",
      "'Electron density statistics': {'5mdi': [{'author_provided': {'r_free_factor': 0.20583, 'resolution_low': 44.76, 'r_factor': 0.16431, 'resolution_high': 2.1, 'percent_r_free': 5}, 'eds_result': {'software_name': 'refmac', 'resolution_low': 44.759, 'r_factor': 0.1636, 'cc_fofc': 0.9734, 'software_version': '5.8.0158 2016/10/03', 'bulk_solvent_b': 47.797, 'diff_map_mean_density': 0, 'map_rmsd_mean_density': 0.16944, 'wilson_scale': 155.60806, 'babinet_bulk_b': 49.137, 'resolution_high': 2.1, 'map_min_density': -0.51846, 'babinet_b': 82.059, 'overall_weighted_f_factor': 0.1465, 'num_reflections': 13158, 'babinet_scale': 0.22, 'completeness': 98.984, 'diff_map_rmsd_mean_density': 0.05224, 'num_reflections_used': 12500, 'coeff': 0.973, 'map_mean_density': 0, 'babinet_bulk_scale': 0.433, 'wilson_b': 40.31, 'diff_map_max_density': 0.35734, 'map_max_density': 2.03579, 'diff_map_min_density': -0.37402}, 'num_strict_non_crystallographic_symmetry_operator': 0, 'cell': {'a': 89.518, 'beta': 90, 'b': 89.518, 'alpha': 90, 'c': 49.153, 'gamma': 120}, 'num_residues': 102, 'space_group': 'P 63'}]}...\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "import csv\n",
    "import time\n",
    "\n",
    "WAIT_TIME = 10\n",
    "TOTAL_TIME = 20000\n",
    "\n",
    "def getDataFromEBI_Electron_density():\n",
    "    # Read input Json file\n",
    "    with open('pdb.txt', 'r') as file:\n",
    "        inputFile = file.read()\n",
    "        print(inputFile)\n",
    "\n",
    "    # the input should contain one or more comma-separated PDB ids.\n",
    "    # API Post URL\n",
    "    # Electron density statistics\n",
    "\n",
    "    post_url_1 = \"https://www.ebi.ac.uk/pdbe/api/pdb/entry/electron_density_statistics/\"\n",
    "\n",
    "    # Make POST request with Json Input body\n",
    "    # WV 190206 - ADDED the 'data =' in below for it to work on Python dictionary\n",
    "\n",
    "    response_1 = requests.post(post_url_1, json = inputFile)\n",
    "    queue_id_1 = str(response_1.json())\n",
    "    resp_1 = \"'Electron density statistics': {}...\".format(queue_id_1)\n",
    "    return resp_1\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    results1 = getDataFromEBI_Electron_density()\n",
    "    # save the results to a json file\n",
    "    json1 = json.dumps(results1)\n",
    "    f = open(\"PDBe_Electron_density.json\", \"w\")\n",
    "    f.write(json1)\n",
    "    f.close()\n",
    "    print(results1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prediction of biophysical characteristics:\n",
    "\n",
    "The second part of this tutorial covers the prediction of biophysical characteristics from protein sequence. This involves sending data to a server, getting back the results, and displaying them. The essential parts of this process are:\n",
    "\n",
    "1. Preparing the input: Converting a protein sequence (e.g. as read from a FASTA file) into an appropriate format\n",
    "2. Using the API: Send the input to the API and get back the results, using two examples (DisoMine and DynaMine)\n",
    "3. Plotting the results: Visualise the predictions to obtain insights into your protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  1. Preparing the input\n",
    "\n",
    "The code below can read a protein sequence from a FASTA file (`read_fasta()`), which is then converted into a JSON for input to the web server (`fasta2json()`). Note that you can input your own sequence by uncommenting the line starting with `#dic = ` and pasting your own sequence identifier and sequence in the Python dictionary. You then have to re-run the code block.\n",
    "\n",
    "The FASTA file used in this example contains the sequence for the p53 protein:\n",
    "\n",
    "```\n",
    ">sp|P04637|P53_HUMAN Cellular tumor antigen p53 OS=Homo sapiens OX=9606 GN=TP53 PE=1 SV=4\n",
    "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGP\n",
    "DEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAK\n",
    "SVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHE\n",
    "RCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNS\n",
    "SCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELP\n",
    "PGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPG\n",
    "GSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def read_fasta(fileName):\n",
    "    dic = {}\n",
    "    cur_scaf = ''\n",
    "    cur_seq = []\n",
    "    for line in open(fileName):\n",
    "        if line.startswith(\">\") and cur_scaf == '':\n",
    "            cur_scaf = line[1:-1].split(' ')[0]\n",
    "        elif line.startswith(\">\") and cur_scaf != '':\n",
    "            dic[cur_scaf] = ''.join(cur_seq)\n",
    "            cur_scaf = line[1:-1].split(' ')[0]\n",
    "            cur_seq = []\n",
    "        else:\n",
    "            cur_seq.append(line.rstrip())\n",
    "            dic[cur_scaf] = ''.join(cur_seq)\n",
    "    return dic\n",
    "\n",
    "def fasta2json(dic):\n",
    "    \"\"\"\n",
    "    Read fasta-format file fil, return dict of form scaffold:sequence.\n",
    "    \"\"\"\n",
    "    with open('output.json', 'w+') as outfile:\n",
    "        json.dump(dic, outfile, indent=4)\n",
    "    return dic\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    dic = read_fasta(\"P53.fasta\")\n",
    "\n",
    "    #dic = {'mySeq': \"ENTERYERSEQENCEHEREPLEASECHANGE\"}\n",
    "    results = fasta2json(dic)\n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Using the API (DisoMine)\n",
    "The following steps are necessary to get the results from the DisoMine REST API:\n",
    "\n",
    "1. Read Json-format file as the input and loads it using the \"json\" module.\n",
    "2. Make POST request with Json input body using \"post\" function.\n",
    "3. Validate the response (Status 202 indicates that request has been accepted for processing)\n",
    "4. Makes GET request to check queue request status.\n",
    "5. Save the results to file when ready"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import time\n",
    "WAIT_TIME = 5\n",
    "TOTAL_TIME = 60\n",
    "def getDataFromDisoMine():\n",
    "\n",
    "    # Read input Json file\n",
    "    file = open('P53.json', 'r')\n",
    "    inputFile = file.read()\n",
    "    request_json = json.dumps(json.loads(inputFile))\n",
    "    request_json = request_json[:-1] + ', \"extra_predictions\":\"true\"}'\n",
    "    \n",
    "    # API Post URL\n",
    "    post_url = \"http://bio2byte.be/disomine/api/\"\n",
    "\n",
    "    # Make POST request with Json Input body\n",
    "    # WV 190206 - ADDED the 'data =' in below for it to work on Python dictionary\n",
    "    response = requests.post(post_url, json = request_json)\n",
    "    \n",
    "    queue_id = str(response.json()['queue_id'])\n",
    "    \n",
    "    # Validating Response Code\n",
    "    assert response.status_code == 202\n",
    "    \n",
    "    print(\"Submitted with queue ID: {}...\".format(queue_id))\n",
    "        \n",
    "    # Now get results - this will happen when results available or when time limit reached\n",
    "    time_waited = 0    \n",
    "    get_url = \"http://bio2byte.be/disomine/api/queue/\"\n",
    "    \n",
    "    while time_waited / TOTAL_TIME < 1.0:\n",
    "\n",
    "      # Get Request\n",
    "      response = requests.get(get_url + queue_id)\n",
    "      \n",
    "      if response.status_code == 202:\n",
    "        print(\"  Waiting for results...\")\n",
    "\n",
    "      elif response.status_code == 200:\n",
    "        queryResults = json.loads(response.text)\n",
    "        break\n",
    "       \n",
    "      else:\n",
    "        print(\"Server error, response {}.\".format(response.status_code))\n",
    "        break\n",
    "      \n",
    "      time.sleep(WAIT_TIME)\n",
    "      time_waited += WAIT_TIME\n",
    "\n",
    "\n",
    "    return queryResults\n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    results = getDataFromDisoMine()\n",
    "\n",
    "    # save the results to a json file\n",
    "    json = json.dumps(results)\n",
    "    f = open(\"Disomine_backbone_sidechain_efoldmine.json\", \"w\")\n",
    "    f.write(json)\n",
    "    f.close()\n",
    "    \n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Using the API (DynaMine)\n",
    "\n",
    "The DynaMine API is older but has additional functionality. It works in a similar way but can take different types of input. The methods of the DynaMineJSONInterface class are: \n",
    "\n",
    "1. A json_api_key and a boolean flag for running status (_running) are defined in the constructor method.\n",
    "2. submit_sequence: The method accepts the protein sequence and a boolean var for prediction status as the inputs and return a discription of the submitted job (job ID).  \n",
    "3. submit_uniprot_ids: The method does the same thing as the submit_sequence but here, accepts the uniprot ID.  \n",
    "4. submit_job: The method accepts the job dict which is just created. It uses the poll_results and  print_progress methods to:\n",
    "* Gives us a job ID.\n",
    "* Makes the call blocking until the results are ready\n",
    "5. dynamine_request: This is the post request with the json encoding of the job. It uses urllib modules  \n",
    "6. poll_results: The job ID is the input and return a discription of the submitted request.  \n",
    "7. print_progress:  It accepts the status and flag as the inputs abd based on  the status of the request ('queued' , 'running' or 'completed' ) it:\n",
    "* Changes the state of the  '_running' boolean.\n",
    "* Gives the status while processing the request.\n",
    "\n",
    "You can also access the [DynaMine server](http://dynamine.ibsquare.be/) directly, where you can also find information on the meaning of the predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Created by Elisa Cilia on 2014-08-20.\n",
    "Copyright (c) 2014 Elisa Cilia. All rights reserved.\n",
    "\"\"\"\n",
    "import json\n",
    "import urllib.request, urllib.parse, urllib.error\n",
    "import time\n",
    "import sys\n",
    "\n",
    "# creating a JSON interface\n",
    "\n",
    "class DynaMineJSONInterface:\n",
    "\n",
    "    def __init__(self, json_api_key):\n",
    "        self._running = False\n",
    "        self._json_api_key = '3469626345366a923473c455fd6bb283621e70a3ba5428f301501f8a'\n",
    "\n",
    "    def submit_sequences(self, proteins, predictions_only=True):\n",
    "        job = {'protocol': '1.0',\n",
    "               'json_api_key': self._json_api_key,\n",
    "               'sequences': proteins,\n",
    "               'predictions_only': predictions_only,\n",
    "               }\n",
    "        return self._submit_job(job)\n",
    "\n",
    "    def submit_uniprot_ids(self, uniprot_ids, predictions_only=True):\n",
    "        job = {'protocol': '1.0',\n",
    "               'json_api_key': self._json_api_key,\n",
    "               'uniprot_ids': uniprot_ids,\n",
    "               'predictions_only': predictions_only,\n",
    "               }\n",
    "        return self._submit_job(job)\n",
    "\n",
    "    def _submit_job(self, job):\n",
    "        # response gives us a job_id\n",
    "        response = self._dynamine_request(job)\n",
    "        if response['status'] == 'error':\n",
    "            sys.stderr.write(response['message'] + '\\n')\n",
    "            sys.exit(1)\n",
    "        job_id = response['job_id']\n",
    "        self._print_progress(response['status'], False)\n",
    "        # this makes the call blocking until the results are ready\n",
    "        while response['status'] != 'completed':\n",
    "            time.sleep(1.5)\n",
    "            response = self._poll_results(job_id)\n",
    "            if response['status'] == 'error':\n",
    "                sys.stderr.write(response['message'] + '\\n')\n",
    "                sys.exit(1)\n",
    "            self._print_progress(response['status'], True)\n",
    "        return response['results']\n",
    "\n",
    "    def _dynamine_request(self, request):\n",
    "        # this is the post request with the json encoding of the job\n",
    "        data = urllib.parse.urlencode({'batch': json.dumps(request)}).encode(\"utf-8\")\n",
    "        url = 'http://dynamine.ibsquare.be/batch_request'\n",
    "        response = ''\n",
    "\n",
    "        req = urllib.request.Request(url, data)\n",
    "        http_response = urllib.request.urlopen(req)\n",
    "        response = json.loads(http_response.read())\n",
    "\n",
    "        #response = {'status': 'error', 'message': 'Unable to communicate with the server.Please verify that %s is up and running, otherwise concact the server administrator (ERROR: %s).'}\n",
    "\n",
    "        return response\n",
    "\n",
    "    def _poll_results(self, job_id):\n",
    "        request = {'protocol': '1.0',\n",
    "                   'json_api_key': self._json_api_key,\n",
    "                   'job_id': job_id\n",
    "                   }\n",
    "        return self._dynamine_request(request)\n",
    "\n",
    "    def _print_progress(self, status, flag):\n",
    "        if status == 'queued':\n",
    "            if flag:\n",
    "                sys.stdout.write('Waiting to be processed.\\n')\n",
    "            else:\n",
    "                sys.stdout.write('The request has been submitted.\\n')\n",
    "            self._running = False\n",
    "        elif status == 'running':\n",
    "            if not self._running:\n",
    "                self._running = True\n",
    "                sys.stdout.write('The request is being processed...')\n",
    "            else:\n",
    "                sys.stdout.write('.')\n",
    "        elif status == 'completed':\n",
    "            if not self._running:\n",
    "                sys.stdout.write('The request is being processed...')\n",
    "            sys.stdout.write('done.\\n')\n",
    "        elif status == 'error':\n",
    "            sys.stderr.write('\\n\\nError while processing the request:\\n\\n')\n",
    "        sys.stdout.flush()\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    json_api_key = 'd03e71ca849d3bc5277fd68e7a0f45124df5be84bf348db65218982f'\n",
    "    dynamine = DynaMineJSONInterface(json_api_key)\n",
    "\n",
    "    # Read input Json file\n",
    "    file = open('output.json', 'r')\n",
    "    inputFile = file.read()\n",
    "    proteins = json.loads(inputFile)\n",
    "    \n",
    "results = dynamine.submit_sequences(proteins, predictions_only=False)\n",
    "print(results)\n",
    "\n",
    "# save the results to a json file\n",
    "json = json.dumps(results)\n",
    "f = open(\"dynamine_output.json\", \"w\")\n",
    "f.write(json)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Plotting the results (for DisoMine, DynaMine and EFoldMine)\n",
    "Here we read back in the output from the DisoMine server, which contains several predictions. The steps in the code are:\n",
    "\n",
    "Load a Json-format file with the input (results saved from DisoMine) using the \"json\" module.\n",
    "Defines an index for each residue.\n",
    "Defines the corresponding scores for disorder (DisoMine), side chain dynamics, backbone dynamics (DynaMine) and early folding (EFoldMine) for each residue.\n",
    "Create plots using the matplotlib library.\n",
    "Note that you might have to install the matplotlib library first - run the code block directly underneath."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python3 -mpip install matplotlib  # install the required package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "\n",
    "# Read input Json file\n",
    "file = open('Disomine_backbone_sidechain_efoldmine.json', 'r')\n",
    "input = file.read()\n",
    "request_json = json.loads(input)\n",
    "\n",
    "# residue index\n",
    "seq = request_json[\"results\"][0]['sequence']\n",
    "#DisoMine Score\n",
    "disomine = request_json[\"results\"][0]['disomine']\n",
    "#backbone Score\n",
    "backbone = request_json[\"results\"][0]['backbone']\n",
    "#sidechain Score\n",
    "sidechain = request_json[\"results\"][0]['sidechain']\n",
    "#efoldmine Score\n",
    "efoldmine = request_json[\"results\"][0]['efoldmine']\n",
    "# creating a plot\n",
    "fig = plt.figure(figsize=(18, 16), dpi= 80, facecolor='w', edgecolor='k')\n",
    "\n",
    "#ploting the DisoMine results\n",
    "axes = fig.add_subplot(411)\n",
    "axes.plot(range(len(seq)), disomine, color=\"b\")\n",
    "axes.set_title(\"DisoMine\")\n",
    "\n",
    "#ploting the Backbone results \n",
    "axes = fig.add_subplot(412)\n",
    "axes.plot(range(len(seq)), backbone, color=\"r\")\n",
    "axes.set_title(\"Backbone\")\n",
    "\n",
    "#ploting the SideChain results\n",
    "axes = fig.add_subplot(413)\n",
    "axes.plot(range(len(seq)), sidechain, color=\"black\")\n",
    "axes.set_title(\"SideChain\")\n",
    "\n",
    "#ploting the EFoldMine results\n",
    "axes = fig.add_subplot(414)\n",
    "axes.plot(range(len(seq)), efoldmine, color=\"g\")\n",
    "axes.set_title(\"EFoldMine\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview prediction of deleteriousness of mutants:\n",
    "The second part of this tutorial covers the prediction of the (likely) effect of single amino acid variants of human proteins on the organism. These can be deleterious (bad) for the human, or have no effect. To do so we are using the DEOGEN2 server, and then visualise the prediction scores. The steps are:\n",
    "\n",
    "1. Get the internal deogen2 sequence id from a uniprot id. seq_uniprot_id2seq_id\n",
    "2. Send the data using a POST request and wait for the results getDataFromDEOGEN2 Note that this function will get all the scores for all possible single amino acid variants in this protein.\n",
    "You can also access the [DEOGEN2 MutaFrame server](http://deogen2.mutaframe.com/) server directly, where you can also find information on the meaning of the predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! pip install bs4  # install the required package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "import re\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# A function for getting sequence id from sequence uniprot ID.\n",
    "# You can enter the uniprot id of your preferred human protein sequence.  \n",
    "def seq_uniprot_id2seq_id(seq_prefix):\n",
    "\n",
    "    # Set the URL to webscrape from\n",
    "    url = 'https://deogen2.mutaframe.com/sequence_finder.php?152539&sequence_prefix='+seq_prefix\n",
    "\n",
    "    # Connect to the URL\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Parse HTML and save to BeautifulSoup objectÂ¶\n",
    "    soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "    p = re.compile(r'data_sequence_id=\"\\d\\d\\d\\d\\d')\n",
    "    mo = p.search(str(soup.contents)).group()\n",
    "    sequence_id = mo[18:]\n",
    "    print(sequence_id)\n",
    "    return sequence_id\n",
    "     \n",
    "# Getting data from DEOGEN2.\n",
    "def getDataFromDEOGEN2(sequence_id, mutation):\n",
    "\n",
    "    # defining the api-endpoint\n",
    "    api_endpoint = \"https://deogen2.mutaframe.com/fetch_results.php\"\n",
    "\n",
    "    # API key\n",
    "    key = 1524214674089\n",
    "    sequence_id = sequence_id\n",
    "    mutation = mutation\n",
    "    structure_id = 0\n",
    "\n",
    "    # data to be sent to api\n",
    "    data = {\n",
    "        \" \": key,\n",
    "        \"sequence_id\": sequence_id,\n",
    "        \"mutation\": mutation,\n",
    "        \"structure_id\": structure_id,\n",
    "    }\n",
    "\n",
    "    # Make POST request with Input body\n",
    "    response = requests.get(api_endpoint, data)\n",
    "    queryResults = json.loads(response.text)\n",
    "    return queryResults\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    seq_uniprot_id = 'P36897'\n",
    "    #seq_uniprot_id = '' # You may uncomment it and write the uniprot id of your preferred human protein.\n",
    "    seq_uniprot_id2seq_id(seq_uniprot_id )\n",
    "    \n",
    "    results = getDataFromDEOGEN2(64689, 'N45S')\n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualising the DEOGEN2 results for a sequence position\n",
    "The steps to visualise the results are:\n",
    "1. Load the JSON file with the results from the DEOGEN2 server for all possible single amino acid variants\n",
    "2. Pull out the scores for a particular sequence position (defined by `res`)\n",
    "3. Create the plots using functions from the matplotlib library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "\n",
    "def deogen2plot(res):\n",
    "\n",
    "    # Read input Json file\n",
    "    file = open('DEOGEN2.json', 'r')\n",
    "    input = file.read()\n",
    "    request_json = json.loads(input)\n",
    "\n",
    "    # mutation DEOGEN2 scores\n",
    "\n",
    "    aa = 'CDSQKIPTFNGHLRWAVEYM'\n",
    "    prediction = []\n",
    "    mutations =[]\n",
    "    for x in aa:\n",
    "        if x != res[0]:\n",
    "            df = request_json['all_deogen2'][res + x]\n",
    "            prediction.append(df)\n",
    "            mutations.append(res + x)\n",
    "\n",
    "    deogen2 = [float(i) for i in prediction]\n",
    "\n",
    "    # creating a plot\n",
    "\n",
    "    fig = plt.figure(figsize=(18, 16), dpi= 80, facecolor='w', edgecolor='k')\n",
    "\n",
    "    #ploting the DEOGEN2 results\n",
    "    \n",
    "    axes = fig.add_subplot(312)\n",
    "    axes.bar(mutations, deogen2)\n",
    "    plt.xticks(mutations, size = 12)\n",
    "    axes.set_title(\"Results Of DEOGEN2 For Amino Acid Replacement Of \" + res , size = 15)\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    results = deogen2plot('R372')\n",
    "    print(results)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
