{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prediction and annotation of protein properties\n",
    "\n",
    "The first part of this tutorial covers the annotation/experimental data available for proteins and the second part covers the prediction of protein properties.\n",
    "1. Experimental data/ annotation\n",
    "2. Prediction\n",
    "\n",
    "<font color='red'>Niloofar: I was wondering whether there might be\n",
    "performance issues using a long jupyter file like this - seems to work\n",
    "OK but not sure about the web version. Might have to split this up, let's\n",
    "check</font>\n",
    "\n",
    "\n",
    "## Experimental data / annotation\n",
    "\n",
    "For only a small fraction of proteins (about 2%) the structure is available. PDBe-KB can be used to to place structural data to its biological context. For the rest (98% of all sequences) we can try to predict the properties from their sequences.\n",
    "\n",
    "\n",
    "## Protein REST API (Uniprot)\n",
    "[Proteins REST API](https://www.ebi.ac.uk/proteins/api/doc/) provides access to key biological data from UniProt and data from Large Scale Studies (LSS) mapped to UniProt.  It has five services (proteins, features, variation, proteomics, antigen). \n",
    "\n",
    "<font color='red'>Niloofar: Maybe better to move the Uniprot API first, so before PDBe-KB</font><font color='blue'>The changes were made</font>\n",
    "\n",
    "Access the API:\n",
    "\n",
    "A) [REST API](https://www.ebi.ac.uk/proteins/api/doc/#/features)\n",
    "\n",
    "B) Python script\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 1: Search by a UniProt accession in the proteins service\n",
    "# You can access other features using the format used in the features page\n",
    "# so for example /proteins/{accession}/isoforms to get isoform information\n",
    "#\n",
    "\n",
    "uniProtId = 'P21802'\n",
    "\n",
    "requestURL = \"https://www.ebi.ac.uk/proteins/api/proteins?offset=0&size=100&accession={}\".format(uniProtId)\n",
    "\n",
    "r = requests.get(requestURL)\n",
    "\n",
    "if not r.ok:\n",
    "    r.raise_for_status()\n",
    "    sys.exit()\n",
    "\n",
    "responseBody = r.text\n",
    "print(responseBody)\n",
    "\n",
    "#\n",
    "# Example 2: Coordinates search using a gene name\n",
    "#\n",
    "\n",
    "geneId = 'FGFR2'\n",
    "\n",
    "requestURL = \"https://www.ebi.ac.uk/proteins/api/coordinates?offset=0&size=100&gene={}\".format(geneId)\n",
    "\n",
    "r = requests.get(requestURL)\n",
    "\n",
    "if not r.ok:\n",
    "    r.raise_for_status()\n",
    "    sys.exit()\n",
    "\n",
    "responseBody = r.text\n",
    "print(responseBody)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PDBe-KB\n",
    "\n",
    "The PDBe-KB (Protein Data Bank in Europe- Knowledge Base) collecs functional annotations and predictions for proteins with structure data in the PDB archive. You can access the [PDBe-KB server](https://www.ebi.ac.uk/pdbe/pdbe-kb) directly. You can try it out by entering a PDB or UniProt identifier.\n",
    "\n",
    "## PDBe REST API\n",
    "\n",
    "The [PDBe REST API](https://www.ebi.ac.uk/pdbe/api/doc/pdb.html) has nine parts (PDB,COMPOUNDS, EMDB,NUCLEIC_MAPPING,SIFTS ,PISA, VALIDATION, TOPOLOGY, SEARCH). They are accessible from different REST API calls. Here are some examples of information the PDB part (first tabs) can give:\n",
    "\n",
    "1.Electron density:\n",
    "The statistics for electron density for structures from X-ray crystallography\n",
    "\n",
    "2.Functional annotation:\n",
    "Functional annotation of all ligands, i.e. 'bound' molecules\n",
    "\n",
    "3.EBI Assembly:\n",
    "Information for each assembly of a given PDB ID (includes the molecule name, type and class, the chains where the molecule occur, and the number of copies of each entity in the assembly)\n",
    "\n",
    "4.EBI Ratio of observed residues:\n",
    "The ratio of observed residues for each chain in each molecule of a pdb entry. The list of chains is sorted by observed_ratio, partial_ratio, and number_residues.\n",
    "\n",
    "5.Binding sites:\n",
    "Binding sites in the entry as per STRUCT_SITE records in PDB files such as ligand, residues in the site, description of the site, etc.\n",
    "\n",
    "6.Secondary structure:\n",
    "Residue ranges of regular secondary structure (alpha helices and beta strands) found in protein chains of the entry. For strands, sheet id can be used to identify a beta sheet.\n",
    "\n",
    "7.Observed ranges:\n",
    "i.e. segments of structural coverage of polymeric molecules that are modelled fully or partly.\n",
    "\n",
    "Access to the API:\n",
    "\n",
    "A) REST API\n",
    "\n",
    "GET for individual queries and POST batch queries. [click here](https://www.ebi.ac.uk/pdbe/api/doc/pdb.html)\n",
    "\n",
    "B) Python script\n",
    "\n",
    "<font color='red'>Niloofar: Can you here get the binding sites? Not many people will know about electon density.\n",
    "I would also prefer that a list of PDB IDs is passed into the function to get the info, not the file\n",
    "read as this adds complication to the code (some people won't know how to program!).</font>\n",
    "<font color ='blue'> The changes were made</font>\n",
    "\n",
    "Below is an example for binding sites. The other properties can be accessed in a similar way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'Binding site': {'5x4f': [], '5mdi': [{'site_id': 'AC1', 'site_residues': [{'entity_id': 1, 'residue_number': 91, 'author_insertion_code': None, 'symmetry_symbol': '1_555', 'chem_comp_id': 'GLY', 'author_residue_number': 69, 'struct_asym_id': 'B', 'chain_id': 'B'}, {'entity_id': 1, 'residue_number': 88, 'author_insertion_code': None, 'symmetry_symbol': '1_555', 'chem_comp_id': 'ALA', 'author_residue_number': 66, 'struct_asym_id': 'B', 'chain_id': 'B'}], 'evidence_code': 'Software', 'details': 'BINDING SITE FOR RESIDUE ACT', 'ligand_residues': [{'entity_id': 2, 'residue_number': 1, 'author_insertion_code': None, 'chain_id': 'B', 'author_residue_number': 101, 'chem_comp_id': 'ACT', 'struct_asym_id': 'C'}]}], '5mrg': []}...\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "import csv\n",
    "\n",
    "def getDataFromBindingSite():\n",
    "    # The input should contain one or more comma-separated PDB ids\n",
    "    input = '2n4p, 5mdi, 5mrg, 5x4f, 6b1g'\n",
    "    # API Post URL\n",
    "    # Binding Site\n",
    "    post_url_1 = \"https://www.ebi.ac.uk/pdbe/api/pdb/entry/binding_sites/\"\n",
    "\n",
    "    # Make POST request\n",
    "    # WV 190206 - ADDED the 'data =' in below for it to work on Python dictionary\n",
    "\n",
    "    response_1 = requests.post(post_url_1, json = input)\n",
    "    queue_id_1 = str(response_1.json())\n",
    "    resp_1 = \"'Binding site': {}...\".format(queue_id_1)\n",
    "    return resp_1\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    results1 = getDataFromBindingSite()\n",
    "    # save the results to a json file\n",
    "    json1 = json.dumps(results1)\n",
    "    f = open(\"BindingSite.json\", \"w\")\n",
    "    f.write(json1)\n",
    "    f.close()\n",
    "    print(results1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## InterProScan\n",
    "[InterProScan](https://www.ebi.ac.uk/interpro/search/sequence-search) uses your input amino acid sequence, in FASTA format,\n",
    "and scans it to find which proteins it matches in terms of structural similarities and functional predictions.\n",
    "\n",
    "A. [API Interface](https://www.ebi.ac.uk/Tools/common/tools/help/index.html?tool=iprscan5)\n",
    "\n",
    "<font color='red'>Niloofar: will you include an example here based on a protein sequence?</font>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STRING\n",
    "This database contains information about protein interactions. Below you can find python code to extract information via the STRING API.\n",
    "\n",
    "<font color='red'>I think for all of these we also have to manually access a web page before moving to the API,\n",
    "so it's more concrete for people, and we can then make the connection to the data via the API</font>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['p53', '0', '9606.ENSP00000269305', '9606', 'Homo sapiens', 'TP53', 'Cellular tumor antigen p53; Acts as a tumor suppressor in many tumor types; induces growth arrest or apoptosis depending on the physiological circumstances and cell type. Involved in cell cycle regulation as a trans-activator that acts to negatively regulate cell division by controlling a set of genes required for this process. One of the activated genes is an inhibitor of cyclin-dependent kinases. Apoptosis induction seems to be mediated either by stimulation of BAX and FAS antigen expression, or by repression of Bcl-2 expression. In cooperation with mitochondrial PPIF is involved in  [...] ']\n",
      "Input:\tp53\tSTRING:\t9606.ENSP00000269305\n",
      "['BRCA1', '1', '9606.ENSP00000418960', '9606', 'Homo sapiens', 'BRCA1', \"Breast cancer type 1 susceptibility protein; E3 ubiquitin-protein ligase that specifically mediates the formation of 'Lys-6'-linked polyubiquitin chains and plays a central role in DNA repair by facilitating cellular responses to DNA damage. It is unclear whether it also mediates the formation of other types of polyubiquitin chains. The E3 ubiquitin-protein ligase activity is required for its tumor suppressor function. The BRCA1-BARD1 heterodimer coordinates a diverse range of cellular pathways such as DNA damage repair, ubiquitination and transcriptional regulation to maintain genomic [...] \"]\n",
      "Input:\tBRCA1\tSTRING:\t9606.ENSP00000418960\n",
      "['cdk2', '2', '9606.ENSP00000266970', '9606', 'Homo sapiens', 'CDK2', 'Cyclin-dependent kinase 2; Serine/threonine-protein kinase involved in the control of the cell cycle; essential for meiosis, but dispensable for mitosis. Phosphorylates CTNNB1, USP37, p53/TP53, NPM1, CDK7, RB1, BRCA2, MYC, NPAT, EZH2. Triggers duplication of centrosomes and DNA. Acts at the G1-S transition to promote the E2F transcriptional program and the initiation of DNA synthesis, and modulates G2 progression; controls the timing of entry into mitosis/meiosis by controlling the subsequent activation of cyclin B/CDK1 by phosphorylation, and coordinates the activation of cyclin B/CDK [...] ']\n",
      "Input:\tcdk2\tSTRING:\t9606.ENSP00000266970\n",
      "['Q99835', '3', '9606.ENSP00000249373', '9606', 'Homo sapiens', 'SMO', \"Smoothened homolog; G protein-coupled receptor that probably associates with the patched protein (PTCH) to transduce the hedgehog's proteins signal. Binding of sonic hedgehog (SHH) to its receptor patched is thought to prevent normal inhibition by patched of smoothened (SMO). Required for the accumulation of KIF7, GLI2 and GLI3 in the cilia. Interacts with DLG5 at the ciliary base to induce the accumulation of KIF7 and GLI2 at the ciliary tip for GLI2 activation (By similarity)\"]\n",
      "Input:\tQ99835\tSTRING:\t9606.ENSP00000249373\n"
     ]
    }
   ],
   "source": [
    "## For a given list of proteins the script resolves them (if possible) to the best \n",
    "## matching STRING identifier and prints out the mapping on screen in the TSV format\n",
    "\n",
    "from __future__ import print_function\n",
    "import sys\n",
    "import requests\n",
    "\n",
    "def string():\n",
    "    string_api_url = \"http://string-db.org/api\"\n",
    "    output_format = \"tsv-no-header\"\n",
    "    method = \"get_string_ids\"\n",
    "\n",
    "    ## construct params dictionary\n",
    "\n",
    "    params = {\n",
    "\n",
    "        \"identifiers\" : \"\\r\".join([\"p53\", \"BRCA1\", \"cdk2\", \"Q99835\"]), # your protein list\n",
    "        \"species\" : 9606, # species NCBI identifier\n",
    "        \"limit\" : 1, # only one (best) identifier per input protein\n",
    "        \"echo_query\" : 1, # see your input identifiers in the output\n",
    "        \"caller_identity\" : \"www.awesome_app.org\" # your app name\n",
    "\n",
    "    }\n",
    "\n",
    "    ## contruct method URL\n",
    "\n",
    "    request_url = string_api_url + \"/\" + output_format + \"/\" + method\n",
    "\n",
    "    ## Call STRING\n",
    "\n",
    "    try:\n",
    "        response = requests.post(request_url, params=params)\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(e)\n",
    "        sys.exit()\n",
    "\n",
    "    ## Read and parse the results\n",
    "\n",
    "    c = 0\n",
    "    with open('string_results.txt', 'w') as f:\n",
    "        f.write(response.text.strip())\n",
    "        for line in response.text.strip().split(\"\\n\"):\n",
    "            l = line.split(\"\\t\")\n",
    "            print(l)\n",
    "            input_identifier, string_identifier = l[0], l[2]\n",
    "\n",
    "            print(\"Input:\", input_identifier, \"STRING:\", string_identifier, sep=\"\\t\")\n",
    "string()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prediction of biophysical characteristics:\n",
    "\n",
    "The second part of this tutorial covers the prediction of biophysical characteristics from protein sequence. This involves sending data to a server, getting back the results, and displaying them. The essential parts of this process are:\n",
    "\n",
    "1. Preparing the input: Converting a protein sequence (e.g. as read from a FASTA file) into an appropriate format\n",
    "2. Using the API: Send the input to the API and get back the results, using two examples (DisoMine and DynaMine)\n",
    "3. Plotting the results: Visualise the predictions to obtain insights into your protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  1. Preparing the input\n",
    "\n",
    "The code below can read a protein sequence from a FASTA file (`read_fasta()`), which is then converted into a JSON for input to the web server (`fasta2json()`). Note that you can input your own sequence by uncommenting the line starting with `#dic = ` and pasting your own sequence identifier and sequence in the Python dictionary. You then have to re-run the code block.\n",
    "\n",
    "The FASTA file used in this example contains the sequence for the p53 protein:\n",
    "\n",
    "```\n",
    ">sp|P04637|P53_HUMAN Cellular tumor antigen p53 OS=Homo sapiens OX=9606 GN=TP53 PE=1 SV=4\n",
    "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGP\n",
    "DEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAK\n",
    "SVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHE\n",
    "RCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNS\n",
    "SCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELP\n",
    "PGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPG\n",
    "GSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'P53.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-12d2505379b1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'__main__'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m     \u001b[0mdic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mread_fasta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"P53.fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m     \u001b[0;31m#dic = {'mySeq': \"ENTERYERSEQENCEHEREPLEASECHANGE\"}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-12d2505379b1>\u001b[0m in \u001b[0;36mread_fasta\u001b[0;34m(fileName)\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mcur_scaf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mcur_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfileName\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\">\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mcur_scaf\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m             \u001b[0mcur_scaf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m' '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'P53.fasta'"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "def read_fasta(fileName):\n",
    "    dic = {}\n",
    "    cur_scaf = ''\n",
    "    cur_seq = []\n",
    "    for line in open(fileName):\n",
    "        if line.startswith(\">\") and cur_scaf == '':\n",
    "            cur_scaf = line[1:-1].split(' ')[0]\n",
    "        elif line.startswith(\">\") and cur_scaf != '':\n",
    "            dic[cur_scaf] = ''.join(cur_seq)\n",
    "            cur_scaf = line[1:-1].split(' ')[0]\n",
    "            cur_seq = []\n",
    "        else:\n",
    "            cur_seq.append(line.rstrip())\n",
    "            dic[cur_scaf] = ''.join(cur_seq)\n",
    "    return dic\n",
    "\n",
    "def fasta2json(dic):\n",
    "    \"\"\"\n",
    "    Read fasta-format file fil, return dict of form scaffold:sequence.\n",
    "    \"\"\"\n",
    "    with open('output.json', 'w+') as outfile:\n",
    "        json.dump(dic, outfile, indent=4)\n",
    "    return dic\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    dic = read_fasta(\"P53.fasta\")\n",
    "\n",
    "    #dic = {'mySeq': \"ENTERYERSEQENCEHEREPLEASECHANGE\"}\n",
    "    results = fasta2json(dic)\n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Using the API (DisoMine)\n",
    "The following steps are necessary to get the results from the DisoMine REST API:\n",
    "\n",
    "1. Read Json-format file as the input and loads it using the \"json\" module.\n",
    "2. Make POST request with Json input body using \"post\" function.\n",
    "3. Validate the response (Status 202 indicates that request has been accepted for processing)\n",
    "4. Makes GET request to check queue request status.\n",
    "5. Save the results to file when ready"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import time\n",
    "WAIT_TIME = 5\n",
    "TOTAL_TIME = 60\n",
    "def getDataFromDisoMine():\n",
    "\n",
    "    # Read input Json file\n",
    "    file = open('P53.json', 'r')\n",
    "    inputFile = file.read()\n",
    "    request_json = json.dumps(json.loads(inputFile))\n",
    "    request_json = request_json[:-1] + ', \"extra_predictions\":\"true\"}'\n",
    "    \n",
    "    # API Post URL\n",
    "    post_url = \"http://bio2byte.be/disomine/api/\"\n",
    "\n",
    "    # Make POST request with Json Input body\n",
    "    # WV 190206 - ADDED the 'data =' in below for it to work on Python dictionary\n",
    "    response = requests.post(post_url, json = request_json)\n",
    "    \n",
    "    queue_id = str(response.json()['queue_id'])\n",
    "    \n",
    "    # Validating Response Code\n",
    "    assert response.status_code == 202\n",
    "    \n",
    "    print(\"Submitted with queue ID: {}...\".format(queue_id))\n",
    "        \n",
    "    # Now get results - this will happen when results available or when time limit reached\n",
    "    time_waited = 0    \n",
    "    get_url = \"http://bio2byte.be/disomine/api/queue/\"\n",
    "    \n",
    "    while time_waited / TOTAL_TIME < 1.0:\n",
    "\n",
    "      # Get Request\n",
    "      response = requests.get(get_url + queue_id)\n",
    "      \n",
    "      if response.status_code == 202:\n",
    "        print(\"  Waiting for results...\")\n",
    "\n",
    "      elif response.status_code == 200:\n",
    "        queryResults = json.loads(response.text)\n",
    "        break\n",
    "       \n",
    "      else:\n",
    "        print(\"Server error, response {}.\".format(response.status_code))\n",
    "        break\n",
    "      \n",
    "      time.sleep(WAIT_TIME)\n",
    "      time_waited += WAIT_TIME\n",
    "\n",
    "\n",
    "    return queryResults\n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    results = getDataFromDisoMine()\n",
    "\n",
    "    # save the results to a json file\n",
    "    json = json.dumps(results)\n",
    "    f = open(\"Disomine_backbone_sidechain_efoldmine.json\", \"w\")\n",
    "    f.write(json)\n",
    "    f.close()\n",
    "    \n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Using the API (DynaMine)\n",
    "\n",
    "The DynaMine API is older but has additional functionality. It works in a similar way but can take different types of input. The methods of the DynaMineJSONInterface class are: \n",
    "\n",
    "1. A json_api_key and a boolean flag for running status (_running) are defined in the constructor method.\n",
    "2. submit_sequence: The method accepts the protein sequence and a boolean var for prediction status as the inputs and return a discription of the submitted job (job ID).  \n",
    "3. submit_uniprot_ids: The method does the same thing as the submit_sequence but here, accepts the uniprot ID.  \n",
    "4. submit_job: The method accepts the job dict which is just created. It uses the poll_results and  print_progress methods to:\n",
    "* Gives us a job ID.\n",
    "* Makes the call blocking until the results are ready\n",
    "5. dynamine_request: This is the post request with the json encoding of the job. It uses urllib modules  \n",
    "6. poll_results: The job ID is the input and return a discription of the submitted request.  \n",
    "7. print_progress:  It accepts the status and flag as the inputs abd based on  the status of the request ('queued' , 'running' or 'completed' ) it:\n",
    "* Changes the state of the  '_running' boolean.\n",
    "* Gives the status while processing the request.\n",
    "\n",
    "You can also access the [DynaMine server](http://dynamine.ibsquare.be/) directly, where you can also find information on the meaning of the predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Created by Elisa Cilia on 2014-08-20.\n",
    "Copyright (c) 2014 Elisa Cilia. All rights reserved.\n",
    "\"\"\"\n",
    "import json\n",
    "import urllib.request, urllib.parse, urllib.error\n",
    "import time\n",
    "import sys\n",
    "\n",
    "# creating a JSON interface\n",
    "\n",
    "class DynaMineJSONInterface:\n",
    "\n",
    "    def __init__(self, json_api_key):\n",
    "        self._running = False\n",
    "        self._json_api_key = '3469626345366a923473c455fd6bb283621e70a3ba5428f301501f8a'\n",
    "\n",
    "    def submit_sequences(self, proteins, predictions_only=True):\n",
    "        job = {'protocol': '1.0',\n",
    "               'json_api_key': self._json_api_key,\n",
    "               'sequences': proteins,\n",
    "               'predictions_only': predictions_only,\n",
    "               }\n",
    "        return self._submit_job(job)\n",
    "\n",
    "    def submit_uniprot_ids(self, uniprot_ids, predictions_only=True):\n",
    "        job = {'protocol': '1.0',\n",
    "               'json_api_key': self._json_api_key,\n",
    "               'uniprot_ids': uniprot_ids,\n",
    "               'predictions_only': predictions_only,\n",
    "               }\n",
    "        return self._submit_job(job)\n",
    "\n",
    "    def _submit_job(self, job):\n",
    "        # response gives us a job_id\n",
    "        response = self._dynamine_request(job)\n",
    "        if response['status'] == 'error':\n",
    "            sys.stderr.write(response['message'] + '\\n')\n",
    "            sys.exit(1)\n",
    "        job_id = response['job_id']\n",
    "        self._print_progress(response['status'], False)\n",
    "        # this makes the call blocking until the results are ready\n",
    "        while response['status'] != 'completed':\n",
    "            time.sleep(1.5)\n",
    "            response = self._poll_results(job_id)\n",
    "            if response['status'] == 'error':\n",
    "                sys.stderr.write(response['message'] + '\\n')\n",
    "                sys.exit(1)\n",
    "            self._print_progress(response['status'], True)\n",
    "        return response['results']\n",
    "\n",
    "    def _dynamine_request(self, request):\n",
    "        # this is the post request with the json encoding of the job\n",
    "        data = urllib.parse.urlencode({'batch': json.dumps(request)}).encode(\"utf-8\")\n",
    "        url = 'http://dynamine.ibsquare.be/batch_request'\n",
    "        response = ''\n",
    "\n",
    "        req = urllib.request.Request(url, data)\n",
    "        http_response = urllib.request.urlopen(req)\n",
    "        response = json.loads(http_response.read())\n",
    "\n",
    "        #response = {'status': 'error', 'message': 'Unable to communicate with the server.Please verify that %s is up and running, otherwise concact the server administrator (ERROR: %s).'}\n",
    "\n",
    "        return response\n",
    "\n",
    "    def _poll_results(self, job_id):\n",
    "        request = {'protocol': '1.0',\n",
    "                   'json_api_key': self._json_api_key,\n",
    "                   'job_id': job_id\n",
    "                   }\n",
    "        return self._dynamine_request(request)\n",
    "\n",
    "    def _print_progress(self, status, flag):\n",
    "        if status == 'queued':\n",
    "            if flag:\n",
    "                sys.stdout.write('Waiting to be processed.\\n')\n",
    "            else:\n",
    "                sys.stdout.write('The request has been submitted.\\n')\n",
    "            self._running = False\n",
    "        elif status == 'running':\n",
    "            if not self._running:\n",
    "                self._running = True\n",
    "                sys.stdout.write('The request is being processed...')\n",
    "            else:\n",
    "                sys.stdout.write('.')\n",
    "        elif status == 'completed':\n",
    "            if not self._running:\n",
    "                sys.stdout.write('The request is being processed...')\n",
    "            sys.stdout.write('done.\\n')\n",
    "        elif status == 'error':\n",
    "            sys.stderr.write('\\n\\nError while processing the request:\\n\\n')\n",
    "        sys.stdout.flush()\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    json_api_key = 'd03e71ca849d3bc5277fd68e7a0f45124df5be84bf348db65218982f'\n",
    "    dynamine = DynaMineJSONInterface(json_api_key)\n",
    "\n",
    "    # Read input Json file\n",
    "    file = open('output.json', 'r')\n",
    "    inputFile = file.read()\n",
    "    proteins = json.loads(inputFile)\n",
    "    \n",
    "results = dynamine.submit_sequences(proteins, predictions_only=False)\n",
    "print(results)\n",
    "\n",
    "# save the results to a json file\n",
    "json = json.dumps(results)\n",
    "f = open(\"dynamine_output.json\", \"w\")\n",
    "f.write(json)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Plotting the results (for DisoMine, DynaMine and EFoldMine)\n",
    "Here we read back in the output from the DisoMine server, which contains several predictions. The steps in the code are:\n",
    "\n",
    "Load a Json-format file with the input (results saved from DisoMine) using the \"json\" module.\n",
    "Defines an index for each residue.\n",
    "Defines the corresponding scores for disorder (DisoMine), side chain dynamics, backbone dynamics (DynaMine) and early folding (EFoldMine) for each residue.\n",
    "Create plots using the matplotlib library.\n",
    "Note that you might have to install the matplotlib library first - run the code block directly underneath."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python3 -mpip install matplotlib  # install the required package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "\n",
    "# Read input Json file\n",
    "file = open('Disomine_backbone_sidechain_efoldmine.json', 'r')\n",
    "input = file.read()\n",
    "request_json = json.loads(input)\n",
    "\n",
    "# residue index\n",
    "seq = request_json[\"results\"][0]['sequence']\n",
    "#DisoMine Score\n",
    "disomine = request_json[\"results\"][0]['disomine']\n",
    "#backbone Score\n",
    "backbone = request_json[\"results\"][0]['backbone']\n",
    "#sidechain Score\n",
    "sidechain = request_json[\"results\"][0]['sidechain']\n",
    "#efoldmine Score\n",
    "efoldmine = request_json[\"results\"][0]['efoldmine']\n",
    "# creating a plot\n",
    "fig = plt.figure(figsize=(18, 16), dpi= 80, facecolor='w', edgecolor='k')\n",
    "\n",
    "#ploting the DisoMine results\n",
    "axes = fig.add_subplot(411)\n",
    "axes.plot(range(len(seq)), disomine, color=\"b\")\n",
    "axes.set_title(\"DisoMine\")\n",
    "\n",
    "#ploting the Backbone results \n",
    "axes = fig.add_subplot(412)\n",
    "axes.plot(range(len(seq)), backbone, color=\"r\")\n",
    "axes.set_title(\"Backbone\")\n",
    "\n",
    "#ploting the SideChain results\n",
    "axes = fig.add_subplot(413)\n",
    "axes.plot(range(len(seq)), sidechain, color=\"black\")\n",
    "axes.set_title(\"SideChain\")\n",
    "\n",
    "#ploting the EFoldMine results\n",
    "axes = fig.add_subplot(414)\n",
    "axes.plot(range(len(seq)), efoldmine, color=\"g\")\n",
    "axes.set_title(\"EFoldMine\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview prediction of deleteriousness of mutants:\n",
    "The second part of this tutorial covers the prediction of the (likely) effect of single amino acid variants of human proteins on the organism. These can be deleterious (bad) for the human, or have no effect. To do so we are using the DEOGEN2 server, and then visualise the prediction scores. The steps are:\n",
    "\n",
    "1. Get the internal deogen2 sequence id from a uniprot id. seq_uniprot_id2seq_id\n",
    "2. Send the data using a POST request and wait for the results getDataFromDEOGEN2 Note that this function will get all the scores for all possible single amino acid variants in this protein.\n",
    "You can also access the [DEOGEN2 MutaFrame server](http://deogen2.mutaframe.com/) server directly, where you can also find information on the meaning of the predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! pip install bs4  # install the required package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "import re\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# A function for getting sequence id from sequence uniprot ID.\n",
    "# You can enter the uniprot id of your preferred human protein sequence.  \n",
    "def seq_uniprot_id2seq_id(seq_prefix):\n",
    "\n",
    "    # Set the URL to webscrape from\n",
    "    url = 'https://deogen2.mutaframe.com/sequence_finder.php?152539&sequence_prefix='+seq_prefix\n",
    "\n",
    "    # Connect to the URL\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Parse HTML and save to BeautifulSoup objectÂ¶\n",
    "    soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "    p = re.compile(r'data_sequence_id=\"\\d\\d\\d\\d\\d')\n",
    "    mo = p.search(str(soup.contents)).group()\n",
    "    sequence_id = mo[18:]\n",
    "    print(sequence_id)\n",
    "    return sequence_id\n",
    "     \n",
    "# Getting data from DEOGEN2.\n",
    "def getDataFromDEOGEN2(sequence_id, mutation):\n",
    "\n",
    "    # defining the api-endpoint\n",
    "    api_endpoint = \"https://deogen2.mutaframe.com/fetch_results.php\"\n",
    "\n",
    "    # API key\n",
    "    key = 1524214674089\n",
    "    sequence_id = sequence_id\n",
    "    mutation = mutation\n",
    "    structure_id = 0\n",
    "\n",
    "    # data to be sent to api\n",
    "    data = {\n",
    "        \" \": key,\n",
    "        \"sequence_id\": sequence_id,\n",
    "        \"mutation\": mutation,\n",
    "        \"structure_id\": structure_id,\n",
    "    }\n",
    "\n",
    "    # Make POST request with Input body\n",
    "    response = requests.get(api_endpoint, data)\n",
    "    queryResults = json.loads(response.text)\n",
    "    return queryResults\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    seq_uniprot_id = 'P36897'\n",
    "    #seq_uniprot_id = '' # You may uncomment it and write the uniprot id of your preferred human protein.\n",
    "    seq_uniprot_id2seq_id(seq_uniprot_id )\n",
    "    \n",
    "    results = getDataFromDEOGEN2(64689, 'N45S')\n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualising the DEOGEN2 results for a sequence position\n",
    "The steps to visualise the results are:\n",
    "1. Load the JSON file with the results from the DEOGEN2 server for all possible single amino acid variants\n",
    "2. Pull out the scores for a particular sequence position (defined by `res`)\n",
    "3. Create the plots using functions from the matplotlib library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "\n",
    "def deogen2plot(res):\n",
    "\n",
    "    # Read input Json file\n",
    "    file = open('DEOGEN2.json', 'r')\n",
    "    input = file.read()\n",
    "    request_json = json.loads(input)\n",
    "\n",
    "    # mutation DEOGEN2 scores\n",
    "\n",
    "    aa = 'CDSQKIPTFNGHLRWAVEYM'\n",
    "    prediction = []\n",
    "    mutations =[]\n",
    "    for x in aa:\n",
    "        if x != res[0]:\n",
    "            df = request_json['all_deogen2'][res + x]\n",
    "            prediction.append(df)\n",
    "            mutations.append(res + x)\n",
    "\n",
    "    deogen2 = [float(i) for i in prediction]\n",
    "\n",
    "    # creating a plot\n",
    "\n",
    "    fig = plt.figure(figsize=(18, 16), dpi= 80, facecolor='w', edgecolor='k')\n",
    "\n",
    "    #ploting the DEOGEN2 results\n",
    "    \n",
    "    axes = fig.add_subplot(312)\n",
    "    axes.bar(mutations, deogen2)\n",
    "    plt.xticks(mutations, size = 12)\n",
    "    axes.set_title(\"Results Of DEOGEN2 For Amino Acid Replacement Of \" + res , size = 15)\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    results = deogen2plot('R372')\n",
    "    print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ELM\n",
    "Below  is a simple python code for extract information from ELM API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "seq= 'MAAGGSGAESAPPTPSMSSLPLAALNVRVRHRLSLFLNVRTQVAADWTGLAEEM\\\n",
    "NFEYLEIRRLETHPDPTRSLLDDWQGRPGASVGRLLELLAKLGRDDVLVELGPS\\\n",
    "IEEDCRKYILKQQQEAAEKPLQVDSVDSSIPWMSGITIRDDPLGQMPEHFDAFI\\\n",
    "CYCPSDIQFVQEMIRQLEQTNYRLKLCVSDRDVLPGTCVWSIASELIEKRCRRM\\\n",
    "VVVVSDDYLQSKECDFQTKFALSLSPGAHQKRLIPVKYKSMKKEFPSILRFITV\\\n",
    "CDYTNPCTKSWFWTRLARALSLP'\n",
    "\n",
    "# defining the api-endpoint\n",
    "api_endpoint = \"http://elm.eu.org/start_search/\"\n",
    "\n",
    "#Submiting a protein identifier\n",
    "def elm_id_submit():\n",
    "    uniprot_id = 'P69905' + '.tsv'\n",
    "    url = api_endpoint + uniprot_id\n",
    "    response = requests.get(url)\n",
    "    # Output file\n",
    "    with open('ELM_result1.txt', 'w') as f:\n",
    "        f.write(response.content.decode(\"utf-8\"))\n",
    "        print(response.content.decode(\"utf-8\"))\n",
    "    return(response.content)\n",
    "\n",
    "#Submitting a single sequence\n",
    "def elm_seq_submit():\n",
    "    url = api_endpoint + seq\n",
    "    response = requests.get(url)\n",
    "    # Output file\n",
    "    with open('ELM_result2.txt', 'w') as f:\n",
    "        f.write(response.content.decode(\"utf-8\"))\n",
    "        print(response.content.decode(\"utf-8\"))\n",
    "    return response.content\n",
    "\n",
    "elm_id_submit()\n",
    "elm_seq_submit()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
